lib/nx_stencil.ex

```diff
@@ -142,46 +183,47 @@ defmodule NxStencil do
 
     send(
       fst,
-      {self(), :start_data,
-       {
-         fst_rank,
-         nil,
-         List.first(inner_pids),
-         nil,
-         Enum.at(data, 0),
-         Enum.at(data, 1),
-         iters
-       }}
+      %InitData{
+        parent: self(),
+        data: %StencilData{
+          up_rank: nil,
+          self_rank: fst_rank,
+          down_rank: List.first(inner_pids),
+          top_halo: nil,
+          self_data: Enum.at(data, 0),
+          bottom_halo: Enum.at(data, 1)
+        },
+        iters: iters
+      }
     )
 
     send(
       lst,
-      {
-        self(),
-        :start_data,
-        {
-          lst_rank,
-          List.last(inner_pids),
-          nil,
-          Enum.at(data, length(data) - 2),
-          Enum.at(data, length(data) - 1),
-          nil,
-          iters
-        }
+      %InitData{
+        parent: self(),
+        data: %StencilData{
+          up_rank: List.last(inner_pids),
+          self_rank: lst_rank,
+          down_rank: nil,
+          top_halo: Enum.at(data, length(data) - 2),
+          self_data: Enum.at(data, length(data) - 1),
+          bottom_halo: nil
+        },
+        iters: iters
       }
     )
 
     IO.puts("sent fst, lst")
 
     IO.puts("Waiting for result")
 
-    for rank <- 0..(proc_count - 1) do
-      receive do
-        {:end_data, ^rank, data} ->
-          IO.puts("from #{rank} -> #{inspect(data)}")
-      after
-        2000 -> IO.puts("Could not receive from rank #{rank}")
-      end
-    end
+    collect(proc_count)
+  end
+
+  def run(out \\ "data_out.json") do
+    {time, data} = :timer.tc(&run_impl/0)
+    IO.puts("Took #{time / 1000_000}s")
+
+    File.write(out, data |> inspect(limit: :infinity))
```

Comment:
Style comment only, it's considered idiomatic to pipe two or more times. In this case you'd use `inspect(data, limit: :infinity)` instead.